From 63948e1b647d0609380cb87eaef882c9363a15b0 Mon Sep 17 00:00:00 2001
From: Oleksandr Suvorov <oleksandr.suvorov@foundries.io>
Date: Mon, 31 Oct 2022 15:46:44 +0200
Subject: [PATCH] [FIO fromlist] gpu: drm: imx: sec_mipi_dsim-imx: fix probe
 exit path

The reset controls of DSIM treat as exclusive, so they should be
released if probing fails. Otherwise, all further probing always
fail and throw a kernel warning [1].

[1]
  [    2.275283] WARNING: CPU: 0 PID: 90 at drivers/reset/core.c:765 __reset_control_get_internal+0x150/0x170
  [    2.298872] Modules linked in:
  [    2.301936] CPU: 0 PID: 90 Comm: kworker/u8:2 Not tainted 5.15.60-lmp-standard #1
  [    2.309430] Hardware name: FSL i.MX8MM EVK board (DT)
  [    2.314489] Workqueue: events_unbound deferred_probe_work_func
  [    2.320336] pstate: 20000005 (nzCv daif -PAN -UAO -TCO -DIT -SSBS BTYPE=--)
  [    2.327306] pc : __reset_control_get_internal+0x150/0x170
  [    2.332713] lr : __of_reset_control_get+0x178/0x1d4
  ...
  [    2.425482] Call trace:
  [    2.425485]  __reset_control_get_internal+0x150/0x170
  [    2.435158]  __of_reset_control_get+0x178/0x1d4
  [    2.435164]  of_reset_control_array_get+0xb8/0x230
* [    2.448397]  imx_sec_dsim_probe+0x23c/0x360
...
  [    2.543033] ------------[ cut here ]------------

Fixes: a31678d546aa1 ("drm/imx: Replace reset flow for DSIM")
Signed-off-by: Oleksandr Suvorov <oleksandr.suvorov@foundries.io>
---

 drivers/gpu/drm/imx/sec_mipi_dsim-imx.c | 12 +++++++++++-
 1 file changed, 11 insertions(+), 1 deletion(-)

diff --git a/drivers/gpu/drm/imx/sec_mipi_dsim-imx.c b/drivers/gpu/drm/imx/sec_mipi_dsim-imx.c
index c870de0b455be..843f8684c3ebe 100644
--- a/drivers/gpu/drm/imx/sec_mipi_dsim-imx.c
+++ b/drivers/gpu/drm/imx/sec_mipi_dsim-imx.c
@@ -456,7 +456,17 @@ static int imx_sec_dsim_probe(struct platform_device *pdev)
 
 	pm_runtime_enable(dev);
 
-	return component_add(dev, &imx_sec_dsim_ops);
+	ret = component_add(dev, &imx_sec_dsim_ops);
+	if (ret)
+		goto error_pm_disable;
+
+	return 0;
+
+error_pm_disable:
+	pm_runtime_disable(dev);
+	sec_dsim_of_put_resets(dsim_dev);
+
+	return ret;
 }
 
 static int imx_sec_dsim_remove(struct platform_device *pdev)
-- 
2.38.1

